src/async_wrap.cc

```diff
@@ -206,22 +207,22 @@ PromiseWrap* PromiseWrap::New(Environment* env,
   obj->SetInternalField(PromiseWrap::kIsChainedPromiseField,
                         parent_wrap != nullptr ? v8::True(env->isolate())
                                                : v8::False(env->isolate()));
-  CHECK_NULL(promise->GetAlignedPointerFromInternalField(0));
-  promise->SetInternalField(0, obj);
+  CHECK_NULL(promise->GetAlignedPointerFromInternalField(BaseObject::kSlot));
+  promise->SetInternalField(BaseObject::kSlot, obj);
   return new PromiseWrap(env, obj, silent);
 }
 
 void PromiseWrap::getIsChainedPromise(Local<String> property,
                                       const PropertyCallbackInfo<Value>& info) {
   info.GetReturnValue().Set(
-    info.Holder()->GetInternalField(kIsChainedPromiseField));
+      info.Holder()->GetInternalField(PromiseWrap::kIsChainedPromiseField));
 }
 
 static PromiseWrap* extractPromiseWrap(Local<Promise> promise) {
-  Local<Value> resource_object_value = promise->GetInternalField(0);
-  if (resource_object_value->IsObject()) {
+  Local<Value> resource_object_value =
+      promise->GetInternalField(BaseObject::kSlot);
+  if (resource_object_value->IsObject())
```

Comment:
I know it's existing logic but if `IsObject() == false`, then what's `resource_object_value` instead?

From quick grepping I infer base_object-inl.h sets it to `this`? It's not safe to call `SetAlignedPointerInInternalField()`, then read it back as a `Local<Value>`.

Reply:
Indeed. I noted that also. There are certainly more than a few oddities in here.

Reply:
In this particular case, the `kSlot` for `PromiseWrap` is set a couple of lines up. Yes it extends from `AsyncWrap` but it appears to not follow the normal behavior -- that is, whereas other `AsyncWrap`/`BaseObject` store an aligned pointer in the slot, `PromiseWrap` stores a `Local<Object>` (set in the `PromiseWrap::New`). The thing that the `PromiseWrap` is storing *should always be a `Local<Promise>`*

Reply:
@jasnell Sorry I didn’t notice this earlier, but: This is setting internals fields on the *Promise* itself, not on a `BaseObject` subclass. It doesn’t make sense to use `BaseObject::kSlot` here and above.

As for Ben’s question, that’s also answered above where the internal field is set in the Promise – it’s a `nullptr`, because that’s how V8 initializes the field by default.

Reply:
Alright, I've revamped the code here to hopefully make it more correct. PTAL